pkgs/development/haskell-modules/generic-builder.nix

```diff
@@ -212,7 +212,14 @@ let
   nativeGhcCommand = "${nativeGhc.targetPrefix}ghc";
 
   buildPkgDb = ghcName: packageConfDir: ''
-    if [ -d "$p/lib/${ghcName}/package.conf.d" ]; then
+    # If this dependency has a package database, then copy the contents of it, unless
+    # it is one of our GHCs. These can appear in our dependencies when we are doing
+    # native builds, and they have  package databases in them, but we do not want to 
+    # copy them over. 
+    # We don't need to, since those packages will be provided by the GHC when we compile 
+    # with it, and doing so can result in having multiple copies of e.g. Cabal in the database 
+    # with the same name and version, which is ambiguous.
+    if [ -d "$p/lib/${ghcName}/package.conf.d" ] && [ "$p" != "${ghc}" ] && [ "$p" != "${nativeGhc}" ]; then
```

Comment:
Have you actually tested this change?

Reply:
Yes, as I mentioned in the description, but to elaborate:
- I tested that the package database for happy's setup compilation (which is where I observed the multiple cabal instances) now has only one, and does what I expect testing with `set -x`.
- I built our packages at work which use the nixpkgs infrastructure with this change and also the GHC patch. That doesn't cover everything, but they use a good subset of the popular hackage packages, so I'm moderately confident that it doesn't break anything. That fit built for both NixOS and macOS.

One caveat is that my testing was based on 18.09, but I don't *think* anything relevant has changed since then.